### PR TITLE
feat(bot): deploy func hosted bot

### PR DESCRIPTION
**Pull request by XiaofuHuang**
Detail: https://msazure.visualstudio.com/Microsoft%20Teams%20Extensibility/_workitems/edit/13776517/

### CODE REVIEW

**Comment by XiaofuHuang** [comment]
E2E TEST: none

File: packages/fx-core/src/plugins/resource/bot/functionsHostedBot/plugin.ts [hunk 1]
```diff
@@ -109,4 +126,105 @@ export class FunctionsHostedBotImpl extends TeamsBotImpl {
...
+      await this.getAzureAccountCredential(),
+      this.config.provision.subscriptionId!
+    );
+    const listResponse = await AzureOperations.ListPublishingCredentials(
+      webSiteMgmtClient,
+      this.config.provision.resourceGroup!,
+      this.config.provision.siteName!
+    );
+
+    const publishingUserName = listResponse.publishingUserName
```
Comment (hund030) [inline packages/fx-core/src/plugins/resource/bot/functionsHostedBot/plugin.ts:195]: NIT: listResponse.publishingUserName ?? ""

**Review by hund030 — APPROVED** [review]

File: packages/fx-core/src/plugins/resource/bot/functionsHostedBot/plugin.ts [hunk 2]
```diff
@@ -109,4 +126,105 @@ export class FunctionsHostedBotImpl extends TeamsBotImpl {
...
+      maxContentLength: Infinity,
+      maxBodyLength: Infinity,
+      timeout: CommonConstants.deployTimeoutInMs,
+    };
+
+    const zipDeployEndpoint: string = getZipDeployEndpoint(this.config.provision.siteName!);
+    await handler?.next(ProgressBarConstants.DEPLOY_STEP_ZIP_DEPLOY);
+    await AzureOperations.ZipDeployPackage(zipDeployEndpoint, zipBuffer, config);
+
+    await AzureOperations.RestartWebApp(
```
Comment (swatDong) [inline packages/fx-core/src/plugins/resource/bot/functionsHostedBot/plugin.ts:218]: Why need restart?

**Review by a1exwang — DISMISSED** [review]

File: packages/fx-core/src/plugins/resource/bot/functionsHostedBot/plugin.ts (hunk 2 above)
Comment (XiaofuHuang) [reply packages/fx-core/src/plugins/resource/bot/functionsHostedBot/plugin.ts:218]: As I discussed with @hund030 offline, the function plugin has met some problems and the solution is to restart.

**Review by swatDong — APPROVED** [review]

File: packages/fx-core/src/plugins/resource/bot/azureOps.ts [hunk 3]
```diff
@@ -179,4 +180,21 @@ export class AzureOperations {
...
+  ): Promise<void> {
+    let res = undefined;
+    try {
+      res = await webSiteMgmtClient.webApps.restart(resourceGroup, siteName);
+    } catch (e) {
+      throw new RestartWebAppError(e);
+    }
+
+    if (!res || !utils.isHttpCodeOkOrCreated(res?._response.status)) {
+      throw new RestartWebAppError();
```
[the remark this answers is not in the document]
Comment (XiaofuHuang) [reply packages/fx-core/src/plugins/resource/bot/azureOps.ts:197]: Good idea, @Siglud @yukun-dong Can I add the status code here? (May be the other azure operation also need to change.)

**Review by a1exwang — APPROVED** [review]

File: packages/fx-core/src/plugins/resource/bot/functionsHostedBot/plugin.ts (hunk 1 above)
Comment (XiaofuHuang) [reply packages/fx-core/src/plugins/resource/bot/functionsHostedBot/plugin.ts]: Thx

File: packages/fx-core/src/plugins/resource/bot/functionsHostedBot/constants.ts [hunk 4]
```diff
@@ -0,0 +1,18 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT license.
+
+import { DeployConfigs } from "../constants";
```
Comment (Siglud) [inline packages/fx-core/src/plugins/resource/bot/functionsHostedBot/constants.ts:4]: seems no use here

File: packages/fx-core/src/plugins/resource/bot/azureOps.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (Siglud) [reply packages/fx-core/src/plugins/resource/bot/azureOps.ts:197]: I think our error class itself is the status code of all the errors. if you mean HTTP response status code,  inner Error will hold that info inside.

File: packages/fx-core/src/plugins/resource/bot/functionsHostedBot/constants.ts (hunk 4 above)
Comment (XiaofuHuang) [reply packages/fx-core/src/plugins/resource/bot/functionsHostedBot/constants.ts]: Removed the useless

**Review by swatDong — APPROVED** [review]

**Review by Siglud — APPROVED** [review]
LGTM

**Pull request merged**